File: app/src/main/java/com/nfcalarmclock/settings/NacAboutSettingFragment.java
package com.nfcalarmclock.settings;

import android.app.Activity;
import android.content.Context;
import android.os.Bundle;
import androidx.preference.Preference;
import com.nfcalarmclock.BuildConfig;
import com.nfcalarmclock.R;
import com.nfcalarmclock.permission.ignorebatteryoptimization.NacIgnoreBatteryOptimizationPermission;
import com.nfcalarmclock.permission.postnotifications.NacPostNotificationsPermission;
import com.nfcalarmclock.permission.readmediaaudio.NacReadMediaAudioPermission;
import com.nfcalarmclock.permission.scheduleexactalarm.NacScheduleExactAlarmPermission;
import com.nfcalarmclock.shared.NacSharedKeys;
import com.nfcalarmclock.util.NacUtility;
import com.nfcalarmclock.whatsnew.NacWhatsNewDialog;

/**
 * Fragment to show the About preferences.
 */
public class NacAboutSettingFragment
	extends NacGenericSettingFragment
{

	/**
	 */
	@Override
	public void onCreatePreferences(Bundle savedInstanceState, String rootKey)
	{
		addPreferencesFromResource(R.xml.about_preferences);

		// Get the version preference
		NacSharedKeys keys = this.getSharedKeys();
		Preference version = findPreference(keys.getVersionPreference());

		// Set the version name as the summary
		version.setSummary(BuildConfig.VERSION_NAME);
	}

	/**
	 * A preference in the tree was clicked.
	 */
	@Override
	public boolean onPreferenceTreeClick(Preference preference)
	{
		Context context = requireContext();
		NacSharedKeys keys = this.getSharedKeys();
		String preferenceKey = preference.getKey();

		// Keys to compare against
		String versionKey = keys.getVersionPreference();
		String disableBatteryOptimizationKey = getString(R.string.about_ignore_battery_optimization_key);
		String nfcKey = getString(R.string.about_nfc_key);
		String storageKey = getString(R.string.about_storage_key);
		String vibrateKey = getString(R.string.about_vibrate_key);
		String foregroundServiceKey = getString(R.string.about_foreground_service_key);
		String fullScreenKey = getString(R.string.about_full_screen_intent_key);
		String scheduleAlarmsKey = getString(R.string.about_schedule_alarm_key);
		String setAlarmKey = getString(R.string.about_set_alarm_key);
		String showNotificationsKey = getString(R.string.about_post_notifications_key);
		String startupKey = getString(R.string.about_boot_key);
		String wakelockKey = getString(R.string.about_wakelock_key);

		// Message if already have this permission
		String message = getString(R.string.message_already_have_permission);

		// Version
		if (preferenceKey.equals(versionKey))
		{
			// Show the What's New dialog
			NacWhatsNewDialog dialog = new NacWhatsNewDialog();

			dialog.show(getChildFragmentManager(), NacWhatsNewDialog.TAG);
		}
		// Disable battery optimization
		else if (preferenceKey.equals(disableBatteryOptimizationKey))
		{
			if (NacIgnoreBatteryOptimizationPermission.hasPermission(context))
			{
				NacUtility.quickToast(context, message);
			}
			else
			{
				Activity activity = requireActivity();

				NacIgnoreBatteryOptimizationPermission.requestPermission(activity);
			}
		}
		// NFC
		else if (preferenceKey.equals(nfcKey))
		{
			NacUtility.quickToast(context, message);
		}
		// Storage
		else if (preferenceKey.equals(storageKey))
		{
			if (NacReadMediaAudioPermission.hasPermission(context))
			{
				NacUtility.quickToast(context, message);
			}
			else
			{
				Activity activity = requireActivity();

				NacReadMediaAudioPermission.requestPermission(activity, 0);
			}
		}
		// Vibrate
		else if (preferenceKey.equals(vibrateKey))
		{
			NacUtility.quickToast(context, message);
		}
		// Foreground service
		else if (preferenceKey.equals(foregroundServiceKey))
		{
			NacUtility.quickToast(context, message);
		}
		// Full screen
		else if (preferenceKey.equals(fullScreenKey))
		{
			NacUtility.quickToast(context, message);
		}
		// Schedule exact alarms
		else if (preferenceKey.equals(scheduleAlarmsKey))
		{
			if (NacScheduleExactAlarmPermission.hasPermission(context))
			{
				NacUtility.quickToast(context, message);
			}
			else
			{
				NacScheduleExactAlarmPermission.requestPermission(requireActivity());
			}
		}
		// Set alarm
		else if (preferenceKey.equals(setAlarmKey))
		{
			NacUtility.quickToast(context, message);
		}
		// Show notifications
		else if (preferenceKey.equals(showNotificationsKey))
		{
			if (NacPostNotificationsPermission.hasPermission(context))
			{
				NacUtility.quickToast(context, message);
			}
			else
			{
				Activity activity = requireActivity();

				NacPostNotificationsPermission.requestPermission(activity, 0);
			}
		}
		// Startup
		else if (preferenceKey.equals(startupKey))
		{
			NacUtility.quickToast(context, message);
		}
		// Wakelock
		else if (preferenceKey.equals(wakelockKey))
		{
			NacUtility.quickToast(context, message);
		}

		// Default return
		return super.onPreferenceTreeClick(preference);
	}

}

File: app/src/main/kotlin/com/nfcalarmclock/settings/NacAboutSettingFragment.kt
package com.nfcalarmclock.settings

import android.app.Activity
import android.os.Build
import android.os.Bundle
import android.view.View
import android.view.ViewGroup
import androidx.core.view.updateLayoutParams
import androidx.preference.Preference
import com.nfcalarmclock.BuildConfig
import com.nfcalarmclock.R
import com.nfcalarmclock.system.permission.ignorebatteryoptimization.NacIgnoreBatteryOptimizationPermission
import com.nfcalarmclock.system.permission.postnotifications.NacPostNotificationsPermission
import com.nfcalarmclock.system.permission.readmediaaudio.NacReadMediaAudioPermission
import com.nfcalarmclock.system.permission.scheduleexactalarm.NacScheduleExactAlarmPermission
import com.nfcalarmclock.system.permission.systemalertwindow.NacSystemAlertWindowPermission
import com.nfcalarmclock.util.NacUtility.quickToast
import com.nfcalarmclock.whatsnew.NacWhatsNewDialog

/**
 * Fragment to show the About preferences.
 */
class NacAboutSettingFragment
	: NacGenericSettingFragment()
{

	/**
	 * Called when creating the preferences.
	 */
	override fun onCreatePreferences(savedInstanceState: Bundle?, rootKey: String?)
	{
		// Inflate the XML file and add the hierarchy to the current preference
		addPreferencesFromResource(R.xml.about_preferences)

		// Get the version preference
		val versionKey = getString(R.string.key_settings_about_version)
		val versionPref = findPreference<Preference>(versionKey)

		// Set the version name as the summary
		versionPref!!.summary = BuildConfig.VERSION_NAME

		// Check if the system alert window permission is needed
		if (!NacSystemAlertWindowPermission.isCorrectAndroidVersion)
		{
			val systemAlertWindowKey = getString(R.string.key_settings_about_system_alert_window)
			val systemAlertWindowPref = findPreference<Preference>(systemAlertWindowKey)

			// Hide the system alert window preference
			systemAlertWindowPref?.isVisible = false
		}
	}

	/**
	 * A preference in the tree was clicked.
	 */
	override fun onPreferenceTreeClick(preference: Preference): Boolean
	{
		val context = requireContext()
		val preferenceKey = preference.key

		// Keys to compare against
		val versionKey = getString(R.string.key_settings_about_version)
		val disableBatteryOptimizationKey =
			getString(R.string.key_settings_about_ignore_battery_optimization)
		val nfcKey = getString(R.string.key_settings_about_nfc)
		val storageKey = getString(R.string.key_settings_about_storage)
		val vibrateKey = getString(R.string.key_settings_about_vibrate)
		val foregroundServiceKey = getString(R.string.key_settings_about_foreground_service)
		val fullScreenKey = getString(R.string.key_settings_about_full_screen_intent)
		val scheduleAlarmsKey = getString(R.string.key_settings_about_schedule_alarm)
		val setAlarmKey = getString(R.string.key_settings_about_set_alarm)
		val showNotificationsKey = getString(R.string.key_settings_about_post_notifications)
		val systemAlertWindowKey = getString(R.string.key_settings_about_system_alert_window)
		val startupKey = getString(R.string.key_settings_about_boot)
		val wakelockKey = getString(R.string.key_settings_about_wakelock)

		// Message if already have this permission
		val messageId = R.string.message_already_have_permission

		// Version
		if (preferenceKey == versionKey)
		{
			// Create the What's New dialog
			val dialog = NacWhatsNewDialog()

			// Show the What's New dialog
			dialog.show(childFragmentManager, NacWhatsNewDialog.TAG)
		}
		// Disable battery optimization
		else if (preferenceKey == disableBatteryOptimizationKey)
		{

			// Show toast to the user saying that the app already has the permission
			if (NacIgnoreBatteryOptimizationPermission.hasPermission(context))
			{
				quickToast(context, messageId)
			}
			// Request the permission
			else
			{
				val activity: Activity = requireActivity()

				NacIgnoreBatteryOptimizationPermission.requestPermission(activity)
			}

		}
		// NFC
		else if (preferenceKey == nfcKey)
		{
			quickToast(context, messageId)
		}
		// Storage
		else if (preferenceKey == storageKey)
		{

			// Show toast to the user saying that the app already has the permission
			if (NacReadMediaAudioPermission.hasPermission(context))
			{
				quickToast(context, messageId)
			}
			// Request the permission
			else
			{
				val activity: Activity = requireActivity()

				NacReadMediaAudioPermission.requestPermission(activity, 0)
			}

		}
		// Vibrate
		else if (preferenceKey == vibrateKey)
		{
			quickToast(context, messageId)
		}
		// Foreground service
		else if (preferenceKey == foregroundServiceKey)
		{
			quickToast(context, messageId)
		}
		// Full screen
		else if (preferenceKey == fullScreenKey)
		{
			quickToast(context, messageId)
		}
		// Schedule exact alarms
		else if (preferenceKey == scheduleAlarmsKey)
		{

			// Show toast to the user saying that the app already has the permission
			if (NacScheduleExactAlarmPermission.hasPermission(context))
			{
				quickToast(context, messageId)
			}
			// Request the permission
			else
			{
				NacScheduleExactAlarmPermission.requestPermission(requireActivity())
			}

		}
		// Set alarm
		else if (preferenceKey == setAlarmKey)
		{
			quickToast(context, messageId)
		}
		// Show notifications
		else if (preferenceKey == showNotificationsKey)
		{

			// Show toast to the user saying that the app already has the permission
			if (NacPostNotificationsPermission.hasPermission(context))
			{
				quickToast(context, messageId)
			}
			// Request the permission
			else
			{
				val activity: Activity = requireActivity()

				NacPostNotificationsPermission.requestPermission(activity, 0)
			}

		}
		// System alert window
		else if (preferenceKey == systemAlertWindowKey)
		{

			// Show toast to the user saying that the app already has the permission
			if (NacSystemAlertWindowPermission.hasPermission(context))
			{
				quickToast(context, messageId)
			}
			// Request the permission
			else
			{
				val activity: Activity = requireActivity()

				NacSystemAlertWindowPermission.requestPermission(activity)
			}

		}
		// Startup
		else if (preferenceKey == startupKey)
		{
			quickToast(context, messageId)
		}
		// Wakelock
		else if (preferenceKey == wakelockKey)
		{
			quickToast(context, messageId)
		}

		// Default return
		return super.onPreferenceTreeClick(preference)
	}

	/**
	 * Called after the view is created.
	 */
	override fun onViewCreated(view: View, savedInstanceState: Bundle?)
	{
		// Setup
		super.onViewCreated(view, savedInstanceState)

		// Check if API < 35, then edge-to-edge is not enforced and do not need to do
		// anything
		if (Build.VERSION.SDK_INT < Build.VERSION_CODES.VANILLA_ICE_CREAM)
		{
			return
		}

		// Setup edge to edge for the recyclerview by using the margin that was saved in
		// the main settings fragment
		listView.updateLayoutParams<ViewGroup.MarginLayoutParams> {
			topMargin = (activity as NacMainSettingActivity).rvTopMargin
		}
	}

}